File: verifier/node/src/artifact/semantic-evaluator.mjs
import fs from "fs";
import path from "path";

const ALLOWED_VERDICTS = new Set([
  "VERIFIED",
  "FAILED",
  "INVALIDATED",
  "INCOMPLETE",
  "NOT_EXECUTABLE"
]);

const ALLOWED_STATUS = new Set([
  "EXECUTED",
  "NOT_EXECUTED"
]);

function readJson(p) {
  return JSON.parse(fs.readFileSync(p, "utf8"));
}

function readTextIfExists(p) {
  return fs.existsSync(p) ? fs.readFileSync(p, "utf8").trim() : null;
}

function stableUnique(items) {
  return [...new Set(items)].sort();
}

function requireArtifactShape(artifact) {
  const requiredTop = ["artifact_id", "title", "subject", "claim_set"];
  for (const k of requiredTop) {
    if (!(k in artifact)) {
      throw new Error(`artifact missing required field: ${k}`);
    }
  }
  if (!artifact.subject || typeof artifact.subject !== "object") {
    throw new Error("artifact subject must be an object");
  }
}

function resolveEvidencePath(repoRoot, artifactPath, declaredPath) {
  if (!declaredPath || typeof declaredPath !== "string") return null;
  if (path.isAbsolute(declaredPath)) return declaredPath;
  const artifactDir = path.dirname(artifactPath);
  const candidateFromArtifact = path.resolve(artifactDir, declaredPath);
  if (fs.existsSync(candidateFromArtifact)) return candidateFromArtifact;
  return path.resolve(repoRoot, declaredPath);
}

function evaluateArtifact0002(repoRoot, artifact, artifactPath) {
  const support = artifact.supporting_evidence || {};

  const sealLog = resolveEvidencePath(repoRoot, artifactPath, support.seal_presence_log);
  const ledgerLog = resolveEvidencePath(repoRoot, artifactPath, support.ledger_presence_log);

  if (!sealLog || !fs.existsSync(sealLog) || !ledgerLog || !fs.existsSync(ledgerLog)) {
    return {
      artifact_id: artifact.artifact_id,
      subject_ref: artifact.subject.subject_ref,
      commit: artifact.subject.commit,
      verdict: "NOT_EXECUTABLE",
      status: "NOT_EXECUTED",
      reason_codes: stableUnique([
        "DECLARED_EVIDENCE_MISSING",
        "EXECUTION_PREREQUISITE_MISSING"
      ]),
      observations: {
        seal_presence: "UNKNOWN",
        ledger_presence: "UNKNOWN"
      },
      notes: [
        "declared mandatory evidence for semantic readiness evaluation was missing"
      ]
    };
  }

  const sealPresence = readTextIfExists(sealLog);
  const ledgerPresence = readTextIfExists(ledgerLog);

  const reasonCodes = ["DECLARED_EVIDENCE_PRESENT"];
  const observations = {
    seal_presence: sealPresence || "UNKNOWN",
    ledger_presence: ledgerPresence || "UNKNOWN"
  };

  if (sealPresence === "ABSENT") reasonCodes.push("GENESIS_SEAL_ABSENT");
  if (sealPresence === "PRESENT") reasonCodes.push("GENESIS_SEAL_PRESENT");
  if (ledgerPresence === "ABSENT") reasonCodes.push("AUTHORITY_LEDGER_ABSENT");
  if (ledgerPresence === "PRESENT") reasonCodes.push("AUTHORITY_LEDGER_PRESENT");

  const readinessUnsatisfied =
    sealPresence === "ABSENT" || ledgerPresence === "ABSENT";

  if (readinessUnsatisfied) {
    reasonCodes.push("SEMANTIC_READINESS_UNSATISFIED");
    return {
      artifact_id: artifact.artifact_id,
      subject_ref: artifact.subject.subject_ref,
      commit: artifact.subject.commit,
      verdict: "FAILED",
      status: "EXECUTED",
      reason_codes: stableUnique(reasonCodes),
      observations,
      notes: [
        "semantic readiness conditions were evaluated from declared evidence"
      ]
    };
  }

  reasonCodes.push("SEMANTIC_VALIDATION_SUCCEEDED");
  return {
    artifact_id: artifact.artifact_id,
    subject_ref: artifact.subject.subject_ref,
    commit: artifact.subject.commit,
    verdict: "VERIFIED",
    status: "EXECUTED",
    reason_codes: stableUnique(reasonCodes),
    observations,
    notes: [
      "semantic readiness conditions were satisfied from declared evidence"
    ]
  };
}

function evaluateArtifact0003(repoRoot, artifact, artifactPath) {
  const support = artifact.supporting_evidence || {};

  const sealLog = resolveEvidencePath(repoRoot, artifactPath, support.seal_presence_log);
  const ledgerLog = resolveEvidencePath(repoRoot, artifactPath, support.ledger_presence_log);

  if (!sealLog || !fs.existsSync(sealLog) || !ledgerLog || !fs.existsSync(ledgerLog)) {
    return {
      artifact_id: artifact.artifact_id,
      subject_ref: artifact.subject.subject_ref,
      commit: artifact.subject.commit,
      verdict: "NOT_EXECUTABLE",
      status: "NOT_EXECUTED",
      reason_codes: stableUnique([
        "DECLARED_EVIDENCE_MISSING",
        "EXECUTION_PREREQUISITE_MISSING"
      ]),
      observations: {
        seal_presence: "UNKNOWN",
        ledger_presence: "UNKNOWN"
      },
      notes: [
        "declared mandatory evidence for issuance validation was missing"
      ]
    };
  }

  const sealPresence = readTextIfExists(sealLog);
  const ledgerPresence = readTextIfExists(ledgerLog);

  const reasonCodes = ["DECLARED_EVIDENCE_PRESENT"];
  const observations = {
    seal_presence: sealPresence || "UNKNOWN",
    ledger_presence: ledgerPresence || "UNKNOWN"
  };

  if (sealPresence === "ABSENT") reasonCodes.push("GENESIS_SEAL_ABSENT");
  if (sealPresence === "PRESENT") reasonCodes.push("GENESIS_SEAL_PRESENT");
  if (ledgerPresence === "ABSENT") reasonCodes.push("AUTHORITY_LEDGER_ABSENT");
  if (ledgerPresence === "PRESENT") reasonCodes.push("AUTHORITY_LEDGER_PRESENT");

  const subjectAbsent =
    sealPresence === "ABSENT" || ledgerPresence === "ABSENT";

  if (subjectAbsent) {
    reasonCodes.push("SUBJECT_ABSENT_NOT_EXECUTABLE");
    return {
      artifact_id: artifact.artifact_id,
      subject_ref: artifact.subject.subject_ref,
      commit: artifact.subject.commit,
      verdict: "NOT_EXECUTABLE",
      status: "NOT_EXECUTED",
      reason_codes: stableUnique(reasonCodes),
      observations,
      notes: [
        "claimed external issuance subject was absent at execution time"
      ]
    };
  }

  reasonCodes.push("SEMANTIC_VALIDATION_SUCCEEDED");
  return {
    artifact_id: artifact.artifact_id,
    subject_ref: artifact.subject.subject_ref,
    commit: artifact.subject.commit,
    verdict: "VERIFIED",
    status: "EXECUTED",
    reason_codes: stableUnique(reasonCodes),
    observations,
    notes: [
      "claimed external issuance subject was present and executable"
    ]
  };
}


function evaluateArtifact0005(repoRoot, artifact, artifactPath) {
  const support = artifact.supporting_evidence || {};

  const executionStatusPath = resolveEvidencePath(repoRoot, artifactPath, support.execution_status);
  const receiptPath = resolveEvidencePath(repoRoot, artifactPath, support.receipt_json);
  const receiptDigestPath = resolveEvidencePath(repoRoot, artifactPath, support.receipt_digest);
  const outputDigestsPath = resolveEvidencePath(repoRoot, artifactPath, support.output_digests);
  const exitCodePath = resolveEvidencePath(repoRoot, artifactPath, support.exit_code_log);

  if (
    !executionStatusPath || !fs.existsSync(executionStatusPath) ||
    !receiptPath || !fs.existsSync(receiptPath) ||
    !receiptDigestPath || !fs.existsSync(receiptDigestPath) ||
    !outputDigestsPath || !fs.existsSync(outputDigestsPath) ||
    !exitCodePath || !fs.existsSync(exitCodePath)
  ) {
    return {
      artifact_id: artifact.artifact_id,
      subject_ref: artifact.subject.subject_ref,
      commit: artifact.subject.commit,
      verdict: "NOT_EXECUTABLE",
      status: "NOT_EXECUTED",
      reason_codes: stableUnique([
        "DECLARED_EVIDENCE_MISSING",
        "EXECUTION_PREREQUISITE_MISSING"
      ]),
      observations: {
        execution_status: "UNKNOWN",
        receipt_status: "UNKNOWN",
        receipt_authority_binding: "UNKNOWN",
        exit_code: "UNKNOWN"
      },
      notes: [
        "declared mandatory evidence for governed execution validation was missing"
      ]
    };
  }

  const executionStatus = readTextIfExists(executionStatusPath) || "";
  const receipt = readJson(receiptPath);
  const exitCode = readTextIfExists(exitCodePath) || "UNKNOWN";

  const expectedSealId = artifact.subject.authority_seal_id || "UNKNOWN";
  const actualSealId = receipt.authority_seal_id || "UNKNOWN";
  const bindingMatched = actualSealId === expectedSealId;

  const reasonCodes = ["DECLARED_EVIDENCE_PRESENT"];
  const observations = {
    execution_status: executionStatus.includes("STATUS: GOVERNED EXECUTION RECORDED") ? "RECORDED" : "UNKNOWN",
    receipt_status: receipt.receipt_id ? "PRESENT" : "ABSENT",
    receipt_authority_binding: bindingMatched ? "MATCHED" : "MISMATCH",
    exit_code: exitCode
  };

  if (observations.execution_status === "RECORDED") reasonCodes.push("GOVERNED_EXECUTION_RECORDED");
  if (observations.receipt_status === "PRESENT") reasonCodes.push("RECEIPT_PRESENT");
  if (bindingMatched) reasonCodes.push("RECEIPT_AUTHORITY_MATCHED");
  if (exitCode === "0") reasonCodes.push("EXIT_CODE_ZERO");

  const failed =
    observations.execution_status !== "RECORDED" ||
    observations.receipt_status !== "PRESENT" ||
    !bindingMatched ||
    exitCode !== "0";

  if (failed) {
    reasonCodes.push("SEMANTIC_VALIDATION_UNSATISFIED");
    return {
      artifact_id: artifact.artifact_id,
      subject_ref: artifact.subject.subject_ref,
      commit: artifact.subject.commit,
      verdict: "FAILED",
      status: "EXECUTED",
      reason_codes: stableUnique(reasonCodes),
      observations,
      notes: [
        "governed execution evidence was present but one or more required validation conditions failed"
      ]
    };
  }

  reasonCodes.push("SEMANTIC_VALIDATION_SUCCEEDED");
  return {
    artifact_id: artifact.artifact_id,
    subject_ref: artifact.subject.subject_ref,
    commit: artifact.subject.commit,
    verdict: "VERIFIED",
    status: "EXECUTED",
    reason_codes: stableUnique(reasonCodes),
    observations,
    notes: [
      "governed execution evidence, receipt presence, authority binding, and exit code all satisfied validation"
    ]
  };
}

function evaluateArtifact(repoRoot, artifact, artifactPath) {
  switch (artifact.artifact_id) {
    case "artifact-0002":
      return evaluateArtifact0002(repoRoot, artifact, artifactPath);
    case "artifact-0003":
      return evaluateArtifact0003(repoRoot, artifact, artifactPath);
    case "artifact-0005":
      return evaluateArtifact0005(repoRoot, artifact, artifactPath);
    default:
      throw new Error(`unsupported artifact id: ${artifact.artifact_id}`);
  }
}

function validateOutput(output) {
  if (!ALLOWED_VERDICTS.has(output.verdict)) {
    throw new Error(`invalid verdict: ${output.verdict}`);
  }
  if (!ALLOWED_STATUS.has(output.status)) {
    throw new Error(`invalid status: ${output.status}`);
  }
  if (!Array.isArray(output.reason_codes)) {
    throw new Error("reason_codes must be an array");
  }
  if (!output.observations || typeof output.observations !== "object") {
    throw new Error("observations must be an object");
  }
}

function main() {
  const artifactArg = process.argv[2];
  if (!artifactArg) {
    throw new Error("usage: node verifier/node/src/artifact/semantic-evaluator.mjs <artifact-json-path>");
  }

  const repoRoot = process.cwd();
  const artifactPath = path.resolve(repoRoot, artifactArg);
  const artifact = readJson(artifactPath);

  requireArtifactShape(artifact);

  const output = evaluateArtifact(repoRoot, artifact, artifactPath);
  validateOutput(output);

  process.stdout.write(JSON.stringify(output, null, 2) + "\n");
}

main();

File: verifier/rust/src/bin/semantic_evaluator.rs
use std::collections::BTreeSet;
use std::env;
use std::fs;
use std::path::{Path, PathBuf};

use serde_json::{json, Value};

fn read_json(path: &Path) -> Value {
    let data = fs::read_to_string(path).expect("failed to read json file");
    serde_json::from_str(&data).expect("failed to parse json")
}

fn read_text_if_exists(path: &Path) -> Option<String> {
    if path.exists() {
        Some(fs::read_to_string(path).expect("failed to read text").trim().to_string())
    } else {
        None
    }
}

fn resolve_evidence_path(repo_root: &Path, artifact_path: &Path, declared: Option<&str>) -> Option<PathBuf> {
    let declared = declared?;
    let declared_path = PathBuf::from(declared);

    if declared_path.is_absolute() {
        return Some(declared_path);
    }

    let artifact_dir = artifact_path.parent().unwrap_or(repo_root);
    let from_artifact = artifact_dir.join(&declared_path);
    if from_artifact.exists() {
        return Some(from_artifact);
    }

    Some(repo_root.join(declared_path))
}

fn sorted_codes(codes: Vec<&str>) -> Vec<String> {
    let mut set = BTreeSet::new();
    for c in codes {
        set.insert(c.to_string());
    }
    set.into_iter().collect()
}

fn support_path<'a>(artifact: &'a Value, key: &str) -> Option<&'a str> {
    artifact.get("supporting_evidence")?.get(key)?.as_str()
}

fn subject_field<'a>(artifact: &'a Value, key: &str) -> Option<&'a str> {
    artifact.get("subject")?.get(key)?.as_str()
}

fn evaluate_artifact_0002(repo_root: &Path, artifact: &Value, artifact_path: &Path) -> Value {
    let seal_log = resolve_evidence_path(repo_root, artifact_path, support_path(artifact, "seal_presence_log"));
    let ledger_log = resolve_evidence_path(repo_root, artifact_path, support_path(artifact, "ledger_presence_log"));

    if !seal_log.as_ref().map(|p| p.exists()).unwrap_or(false)
        || !ledger_log.as_ref().map(|p| p.exists()).unwrap_or(false)
    {
        return json!({
            "artifact_id": artifact["artifact_id"],
            "subject_ref": subject_field(artifact, "subject_ref").unwrap_or(""),
            "commit": subject_field(artifact, "commit").unwrap_or(""),
            "verdict": "NOT_EXECUTABLE",
            "status": "NOT_EXECUTED",
            "reason_codes": sorted_codes(vec![
                "DECLARED_EVIDENCE_MISSING",
                "EXECUTION_PREREQUISITE_MISSING"
            ]),
            "observations": {
                "seal_presence": "UNKNOWN",
                "ledger_presence": "UNKNOWN"
            },
            "notes": [
                "declared mandatory evidence for semantic readiness evaluation was missing"
            ]
        });
    }

    let seal_presence = read_text_if_exists(&seal_log.unwrap()).unwrap_or_else(|| "UNKNOWN".to_string());
    let ledger_presence = read_text_if_exists(&ledger_log.unwrap()).unwrap_or_else(|| "UNKNOWN".to_string());

    let mut codes = vec!["DECLARED_EVIDENCE_PRESENT"];
    if seal_presence == "ABSENT" {
        codes.push("GENESIS_SEAL_ABSENT");
    }
    if seal_presence == "PRESENT" {
        codes.push("GENESIS_SEAL_PRESENT");
    }
    if ledger_presence == "ABSENT" {
        codes.push("AUTHORITY_LEDGER_ABSENT");
    }
    if ledger_presence == "PRESENT" {
        codes.push("AUTHORITY_LEDGER_PRESENT");
    }

    if seal_presence == "ABSENT" || ledger_presence == "ABSENT" {
        codes.push("SEMANTIC_READINESS_UNSATISFIED");
        json!({
            "artifact_id": artifact["artifact_id"],
            "subject_ref": subject_field(artifact, "subject_ref").unwrap_or(""),
            "commit": subject_field(artifact, "commit").unwrap_or(""),
            "verdict": "FAILED",
            "status": "EXECUTED",
            "reason_codes": sorted_codes(codes),
            "observations": {
                "seal_presence": seal_presence,
                "ledger_presence": ledger_presence
            },
            "notes": [
                "semantic readiness conditions were evaluated from declared evidence"
            ]
        })
    } else {
        codes.push("SEMANTIC_VALIDATION_SUCCEEDED");
        json!({
            "artifact_id": artifact["artifact_id"],
            "subject_ref": subject_field(artifact, "subject_ref").unwrap_or(""),
            "commit": subject_field(artifact, "commit").unwrap_or(""),
            "verdict": "VERIFIED",
            "status": "EXECUTED",
            "reason_codes": sorted_codes(codes),
            "observations": {
                "seal_presence": seal_presence,
                "ledger_presence": ledger_presence
            },
            "notes": [
                "semantic readiness conditions were satisfied from declared evidence"
            ]
        })
    }
}

fn evaluate_artifact_0003(repo_root: &Path, artifact: &Value, artifact_path: &Path) -> Value {
    let seal_log = resolve_evidence_path(repo_root, artifact_path, support_path(artifact, "seal_presence_log"));
    let ledger_log = resolve_evidence_path(repo_root, artifact_path, support_path(artifact, "ledger_presence_log"));

    if !seal_log.as_ref().map(|p| p.exists()).unwrap_or(false)
        || !ledger_log.as_ref().map(|p| p.exists()).unwrap_or(false)
    {
        return json!({
            "artifact_id": artifact["artifact_id"],
            "subject_ref": subject_field(artifact, "subject_ref").unwrap_or(""),
            "commit": subject_field(artifact, "commit").unwrap_or(""),
            "verdict": "NOT_EXECUTABLE",
            "status": "NOT_EXECUTED",
            "reason_codes": sorted_codes(vec![
                "DECLARED_EVIDENCE_MISSING",
                "EXECUTION_PREREQUISITE_MISSING"
            ]),
            "observations": {
                "seal_presence": "UNKNOWN",
                "ledger_presence": "UNKNOWN"
            },
            "notes": [
                "declared mandatory evidence for issuance validation was missing"
            ]
        });
    }

    let seal_presence = read_text_if_exists(&seal_log.unwrap()).unwrap_or_else(|| "UNKNOWN".to_string());
    let ledger_presence = read_text_if_exists(&ledger_log.unwrap()).unwrap_or_else(|| "UNKNOWN".to_string());

    let mut codes = vec!["DECLARED_EVIDENCE_PRESENT"];
    if seal_presence == "ABSENT" {
        codes.push("GENESIS_SEAL_ABSENT");
    }
    if seal_presence == "PRESENT" {
        codes.push("GENESIS_SEAL_PRESENT");
    }
    if ledger_presence == "ABSENT" {
        codes.push("AUTHORITY_LEDGER_ABSENT");
    }
    if ledger_presence == "PRESENT" {
        codes.push("AUTHORITY_LEDGER_PRESENT");
    }

    if seal_presence == "ABSENT" || ledger_presence == "ABSENT" {
        codes.push("SUBJECT_ABSENT_NOT_EXECUTABLE");
        json!({
            "artifact_id": artifact["artifact_id"],
            "subject_ref": subject_field(artifact, "subject_ref").unwrap_or(""),
            "commit": subject_field(artifact, "commit").unwrap_or(""),
            "verdict": "NOT_EXECUTABLE",
            "status": "NOT_EXECUTED",
            "reason_codes": sorted_codes(codes),
            "observations": {
                "seal_presence": seal_presence,
                "ledger_presence": ledger_presence
            },
            "notes": [
                "claimed external issuance subject was absent at execution time"
            ]
        })
    } else {
        codes.push("SEMANTIC_VALIDATION_SUCCEEDED");
        json!({
            "artifact_id": artifact["artifact_id"],
            "subject_ref": subject_field(artifact, "subject_ref").unwrap_or(""),
            "commit": subject_field(artifact, "commit").unwrap_or(""),
            "verdict": "VERIFIED",
            "status": "EXECUTED",
            "reason_codes": sorted_codes(codes),
            "observations": {
                "seal_presence": seal_presence,
                "ledger_presence": ledger_presence
            },
            "notes": [
                "claimed external issuance subject was present and executable"
            ]
        })
    }
}


fn evaluate_artifact_0005(repo_root: &Path, artifact: &Value, artifact_path: &Path) -> Value {
    let execution_status_path = resolve_evidence_path(repo_root, artifact_path, support_path(artifact, "execution_status"));
    let receipt_path = resolve_evidence_path(repo_root, artifact_path, support_path(artifact, "receipt_json"));
    let receipt_digest_path = resolve_evidence_path(repo_root, artifact_path, support_path(artifact, "receipt_digest"));
    let output_digests_path = resolve_evidence_path(repo_root, artifact_path, support_path(artifact, "output_digests"));
    let exit_code_path = resolve_evidence_path(repo_root, artifact_path, support_path(artifact, "exit_code_log"));

    if !execution_status_path.as_ref().map(|p| p.exists()).unwrap_or(false)
        || !receipt_path.as_ref().map(|p| p.exists()).unwrap_or(false)
        || !receipt_digest_path.as_ref().map(|p| p.exists()).unwrap_or(false)
        || !output_digests_path.as_ref().map(|p| p.exists()).unwrap_or(false)
        || !exit_code_path.as_ref().map(|p| p.exists()).unwrap_or(false)
    {
        return json!({
            "artifact_id": artifact["artifact_id"],
            "subject_ref": subject_field(artifact, "subject_ref").unwrap_or(""),
            "commit": subject_field(artifact, "commit").unwrap_or(""),
            "verdict": "NOT_EXECUTABLE",
            "status": "NOT_EXECUTED",
            "reason_codes": sorted_codes(vec![
                "DECLARED_EVIDENCE_MISSING",
                "EXECUTION_PREREQUISITE_MISSING"
            ]),
            "observations": {
                "execution_status": "UNKNOWN",
                "receipt_status": "UNKNOWN",
                "receipt_authority_binding": "UNKNOWN",
                "exit_code": "UNKNOWN"
            },
            "notes": [
                "declared mandatory evidence for governed execution validation was missing"
            ]
        });
    }

    let execution_status = read_text_if_exists(&execution_status_path.unwrap()).unwrap_or_else(|| "".to_string());
    let receipt = read_json(&receipt_path.unwrap());
    let exit_code = read_text_if_exists(&exit_code_path.unwrap()).unwrap_or_else(|| "UNKNOWN".to_string());

    let expected_seal_id = artifact["subject"]["authority_seal_id"].as_str().unwrap_or("UNKNOWN");
    let actual_seal_id = receipt["authority_seal_id"].as_str().unwrap_or("UNKNOWN");
    let binding_matched = actual_seal_id == expected_seal_id;

    let receipt_present = receipt.get("receipt_id").is_some() && !receipt["receipt_id"].is_null();

    let mut codes = vec!["DECLARED_EVIDENCE_PRESENT"];
    if execution_status.contains("STATUS: GOVERNED EXECUTION RECORDED") {
        codes.push("GOVERNED_EXECUTION_RECORDED");
    }
    if receipt_present {
        codes.push("RECEIPT_PRESENT");
    }
    if binding_matched {
        codes.push("RECEIPT_AUTHORITY_MATCHED");
    }
    if exit_code == "0" {
        codes.push("EXIT_CODE_ZERO");
    }

    let observations = json!({
        "execution_status": if execution_status.contains("STATUS: GOVERNED EXECUTION RECORDED") { "RECORDED" } else { "UNKNOWN" },
        "receipt_status": if receipt_present { "PRESENT" } else { "ABSENT" },
        "receipt_authority_binding": if binding_matched { "MATCHED" } else { "MISMATCH" },
        "exit_code": exit_code
    });

    let failed =
        !execution_status.contains("STATUS: GOVERNED EXECUTION RECORDED") ||
        !receipt_present ||
        !binding_matched ||
        exit_code != "0";

    if failed {
        codes.push("SEMANTIC_VALIDATION_UNSATISFIED");
        json!({
            "artifact_id": artifact["artifact_id"],
            "subject_ref": subject_field(artifact, "subject_ref").unwrap_or(""),
            "commit": subject_field(artifact, "commit").unwrap_or(""),
            "verdict": "FAILED",
            "status": "EXECUTED",
            "reason_codes": sorted_codes(codes),
            "observations": observations,
            "notes": [
                "governed execution evidence was present but one or more required validation conditions failed"
            ]
        })
    } else {
        codes.push("SEMANTIC_VALIDATION_SUCCEEDED");
        json!({
            "artifact_id": artifact["artifact_id"],
            "subject_ref": subject_field(artifact, "subject_ref").unwrap_or(""),
            "commit": subject_field(artifact, "commit").unwrap_or(""),
            "verdict": "VERIFIED",
            "status": "EXECUTED",
            "reason_codes": sorted_codes(codes),
            "observations": observations,
            "notes": [
                "governed execution evidence, receipt presence, authority binding, and exit code all satisfied validation"
            ]
        })
    }
}

fn main() {
    let repo_root = env::current_dir().expect("failed to get cwd");
    let artifact_arg = env::args().nth(1).expect("usage: semantic_evaluator <artifact-json-path>");
    let artifact_path = repo_root.join(artifact_arg);
    let artifact = read_json(&artifact_path);

    let artifact_id = artifact["artifact_id"].as_str().expect("artifact_id missing");
    let output = match artifact_id {
        "artifact-0002" => evaluate_artifact_0002(&repo_root, &artifact, &artifact_path),
        "artifact-0003" => evaluate_artifact_0003(&repo_root, &artifact, &artifact_path),
        "artifact-0005" => evaluate_artifact_0005(&repo_root, &artifact, &artifact_path),
        _ => panic!("unsupported artifact id"),
    };

    println!("{}", serde_json::to_string_pretty(&output).unwrap());
}